pkg/metrics/observer/query_observer.go

```diff
@@ -0,0 +1,141 @@
+}
+
+// NewPrometheusMetricObserver allows to continuously observe particular Prometheus query result value and react to it.
+// The name should be unique name for this observer.
+// The queryHandlers represents list of Prometheus queries and handlers where the result value and type are passed.
+// The clients are needed to initialize that prometheus API client.
+func NewPrometheusMetricObserver(name string, queryHandlers []Handler, serviceClient corev1client.ServicesGetter, secretsClient corev1client.SecretsGetter, configsClient corev1client.ConfigMapsGetter,
+	routeClient routev1client.RoutesGetter, recorder events.Recorder, operatorClient v1helpers.StaticPodOperatorClient) factory.Controller {
+	observer := &prometheusMetricObserver{
+		prometheusClientFn: metricsclient.NewPrometheusClient,
+		queryHandlers:      queryHandlers,
+		operatorClient:     operatorClient,
+		serviceClient:      serviceClient,
+		secretsClient:      secretsClient,
+		configsClient:      configsClient,
+		routeClient:        routeClient,
+	}
+	return factory.New().ResyncEvery(defaultPollingDuration).WithSync(observer.sync).ToController(name, recorder.WithComponentSuffix(name))
+}
+
+func (o *prometheusMetricObserver) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	var (
+		errors          []error
+		foundConditions []operatorv1.OperatorCondition
+	)
+
+	// list condition types we already know about
+	degradedConditionNames := sets.NewString("PrometheusObserverDegraded")
+
+	prometheusClient, prometheusClientErr := o.prometheusClientFn(o.serviceClient, o.secretsClient, o.configsClient, o.routeClient)
+	// do not error out if we can't get client, but report all handler degraded status as "unknown".
+	if prometheusClientErr != nil {
+		klog.Infof("Failed to get prometheus client: %v", prometheusClientErr)
+	}
+
+	for _, handler := range o.queryHandlers {
+		handlerConditionType := fmt.Sprintf("%s_PrometheusObserverDegraded", handler.Name)
+		// register handler name as operator degraded condition
+		degradedConditionNames.Insert(handlerConditionType)
+		// if we were not able to acquire prometheus client (service does not exists yet, upgrade, etc.) lets report 'unknown' for this condition
+		if prometheusClient == nil {
+			foundConditions = append(foundConditions, operatorv1.OperatorCondition{
+				Type:    handlerConditionType,
+				Status:  operatorv1.ConditionUnknown,
+				Reason:  "PrometheusUnavailable",
+				Message: fmt.Sprintf("Prometheus client is not available: %v", prometheusClientErr),
+			})
+			continue
+		}
+
+		// pass the polling interval to handler so callers know how often we execute this handler
+		handlerCtx := context.WithValue(ctx, "polling_interval", defaultPollingDuration)
+
+		if err := handler.Handler(handlerCtx, controllerContext.Recorder(), prometheusClient); err != nil {
```

Comment:
one handler per query or one handler per controller?  I guess higher order logic could be created with coordinated handlers and we're likely to use this sparingly enough that we only go degraded for individually named metrics.  `EtcdDiskPerformanceDegraded` as opposed to `EtcdMetricsDegraded`.  or `KubeAPIServerFailoverCapacityDegraded` and not `KubeAPIServerMetricsDegraded`.  

Now that I've typed the examples, this seems fine.